cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -207,31 +207,52 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
   }
 
   public void restartAllServiceInstances(HttpRequest request, HttpResponder responder, String serviceName) {
-    restartInstances(responder, serviceName, -1);
+    restartInstances(responder, serviceName, -1, true);
   }
 
   public void restartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName,
                                      int instanceId) {
-    restartInstances(responder, serviceName, instanceId);
+    restartInstances(responder, serviceName, instanceId, false);
   }
 
-  private void restartInstances(HttpResponder responder, String serviceName, int instanceId) {
+  private void restartInstances(HttpResponder responder, String serviceName, int instanceId, boolean restartAll) {
     long startTimeMs = System.currentTimeMillis();
     boolean isSuccess = true;
+    if (!serviceManagementMap.containsKey(serviceName)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+      return;
+    }
+    MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+
     try {
-      if (!serviceManagementMap.containsKey(serviceName)) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
-        return;
+      if (!masterServiceManager.isServiceEnabled()) {
+        throw new IllegalStateException();
```

Comment:
Should this translate into 403 or 409? If it should be 409, then you can just throw a ``ConflictException`` from here. If 403, then you may want to add a ``ForbiddenException`` and a clause for it in ``HttpExceptionHandler`` (its missing right now).